engine/src/main/scala/cromwell/engine/workflow/tokens/TokenQueue.scala

```diff
@@ -31,23 +34,41 @@ final case class TokenQueue(queue: Queue[ActorRef], private [tokens] val pool: T
     *
     * @return the new token queue
     */
-  def enqueue(actor: ActorRef): TokenQueue = {
-    copy(queue = queue.enqueue(actor))
+  def enqueue(placeholder: TokenQueuePlaceholder): TokenQueue = {
+    copy(queue = queue.enqueue(placeholder))
   }
 
   /**
-    * Returns a DequeuedActor if there's an element available, None otherwise.
+    * Returns a dequeue'd actor if one exists and there's a token available for it
+    * Returns an updated token queue based on this request (successful queuees get removed, hogs get sent to the back)
     */
-  def dequeueOption: Option[DequeuedActor] = {
-    for {
-      actorAndNewQueue <- queue.dequeueOption
-      (actor, newQueue) = actorAndNewQueue
-      lease <- pool.tryAcquire()
-    } yield DequeuedActor(LeasedActor(actor, lease), copy(queue = newQueue))
+  def dequeue: DequeueResult = {
+    queue.dequeueOption match {
+      case Some(actorAndNewQueue) =>
+        val (placeholder, newQueue) = actorAndNewQueue
+        val poolAcquisitionResult = pool.tryAcquire(placeholder.hogGroup)
+        shuffleQueueAndGetLease(poolAcquisitionResult, placeholder, queue, newQueue) match {
+          case ShuffledQueueAndLease(Some(lease), nq) => DequeueResult(Option(LeasedActor(placeholder.actor, lease)), copy(queue = nq))
+          case ShuffledQueueAndLease(None, nq) => DequeueResult(None, copy(queue = nq))
+        }
+      case None =>
+        DequeueResult(None, this)
+    }
+  }
+
+  private def shuffleQueueAndGetLease(result: UnhoggableTokenPoolResult,
+                              thisPlaceholder: TokenQueuePlaceholder,
+                              oldQueue: Queue[TokenQueuePlaceholder],
+                              newQueue: Queue[TokenQueuePlaceholder]): ShuffledQueueAndLease = result match {
+    case thl: TokenHoggingLease => ShuffledQueueAndLease(Some(thl), newQueue)
+    case ComeBackLater => ShuffledQueueAndLease(None, oldQueue)
+    // Hog response - move this placeholder to the end of the queue!
+    case Oink => ShuffledQueueAndLease(None, newQueue :+ thisPlaceholder )
   }
 
   /**
     * Returns true if there's at least on element that can be dequeued, false otherwise.
     */
-  def available: Boolean = queue.nonEmpty && pool.leased() < pool.capacity
+  def available: Boolean =
+    queue.nonEmpty && queue.map(_.hogGroup).iterator.exists(hg => pool.available(hg))
```

Comment:
I have a vague reminiscence of something like this being very slow before, because having to loop through (potentially) the entire queue to determine if an element is available. Maybe a `toSet` would be help so we only check for unique hog groups.
If it's possible to have some sort of "perf unit test" with a large queue (maybe 100k) and the only available hog group at the end of the queue it could be interesting to see if it has any impact when trying to dequeue.